### PR TITLE
fix: consensus metric

### PR DESCRIPTION
**Pull request by tty47**
hello team!

Some context about this PR, Torch generates a metric to show how long the chain is running, but sometimes if the service is not available for any reason (*for example, when autodevops creates a new env, torch starts before the nodes*), it generates an issue.

This PR manages this situation creating a task in the background and checking if the end service is available every 10".

Any feedback is welcome!

Cheers! 🚀 

cc: @celestiaorg/devops 
closes: https://github.com/celestiaorg/devops/issues/571

### CODE REVIEW

**Review by smuu — APPROVED** [review]
lgtm

File: pkg/http/server.go [hunk 1]
```diff
@@ -81,6 +78,26 @@ func Run(cfg config.MutualPeersConfig) {
...
+					// The metric was successfully generated, stop the retries.
+					break
+				}
+
+				// Wait for the retry interval before the next execution
+				time.Sleep(retryInterval)
+			}
+		}()
+	}
+
```
Comment (Bidon15) [inline pkg/http/server.go:100]: It's better to separate this into dedicated functions
In addition they need to return errors and we need to handle them properly

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -99,17 +107,17 @@ func WithMetricsBlockHeight(blockHeight, earliestBlockTime, serviceName, namespa
 }
 
 // CalculateDaysDifference based on the date received, returns the number of days since this day.
-func CalculateDaysDifference(inputTimeString string) int {
+func CalculateDaysDifference(inputTimeString string) (int, error) {
```
Comment (Bidon15) [inline pkg/metrics/metrics.go:110]: You technically don't need this as a public func isn;t it?

**Review by Bidon15 — CHANGES_REQUESTED** [review]

File: pkg/http/server.go [hunk 3]
```diff
@@ -127,23 +127,76 @@ func Run(cfg config.MutualPeersConfig) {
...
+		if err == nil {
+			log.Info("Metric generated for the first block...")
+			// The metric was successfully generated, stop the retries.
+			done <- nil
+			return
 		}
+
+		// Wait for the retry interval before the next execution
+		time.Sleep(retryInterval)
 	}
```
Comment (Bidon15) [inline pkg/http/server.go:174]: you need to be more precise with handling not expected behaviour here. I.e. we need to retry if we fail, hence it makes sense to put into the else. In addition, do we plan to watch until forever? If no, it's much better to set a clear deadline to this process (can be like 4-5 min) where you pass the context to a watcher function and then exit the for loop gracefully.

Moreover, I highly recommend to go with select here, if you opt for the option I provided above

File: pkg/http/server.go [hunk 4]
```diff
@@ -127,23 +127,76 @@ func Run(cfg config.MutualPeersConfig) {
...
-			log.Errorf("Error registering metric block_height_1: %v", err)
-			return true
+		log.Info("Initializing goroutine to generate the metric: hash ")
+		// Initialise the goroutine to generate the metric in the background, only if we specify the node in the config.
+		go func() {
+			log.Info("Consensus node defined to get the first block")
+
+			// Watch for events on the watcher channel
+			done := make(chan error)
+			go WatchHashMetric(cfg, done)
```
Comment (Bidon15) [inline pkg/http/server.go:144]: better use errGo for this and have the WatchHashMetric returning error.

https://medium.com/@yardenlaif/go-sync-or-go-home-errgroup-f91a0ee72d3f

**Review by Bidon15 — CHANGES_REQUESTED** [review]
Nice try! Another round to make it good 🥳

File: pkg/http/server.go [hunk 5]
```diff
@@ -127,23 +129,102 @@ func Run(cfg config.MutualPeersConfig) {
...
+				timer := time.NewTimer(retryInterval)
+				select {
+				case <-ctx.Done():
+					// Context canceled, stop the process
+					log.Info("Context canceled, stopping WatchHashMetric.")
+					timer.Stop()
+					return ctx.Err()
+				case <-timer.C:
+					// Continue to the next iteration
+				}
```
Comment (Bidon15) [inline pkg/http/server.go:196]: [nit]: can be a separate private func to keep it neat 
Can be parked as an issue to move on

**Review by Bidon15 — APPROVED** [review]
1 nit, apart from that lgtm

**Pull request merged**